### DISCUSSION TITLE
Is it possible to set strip color based on htmlcolor name

### DISCUSSION THREAD

**Discussion by patfelst** [Makuna/NeoPixelBus#852] [Q&A] [answered]
Something like this:
```c++
strip.SetPixelColor(0, HtmlColor(c_HtmlNameFuchsia));
```
where the html color names are defined in https://github.com/Makuna/NeoPixelBus/blob/master/src/internal/colors/HtmlColorNameStrings.cpp

I read this page but couldn't see an obvious answer.
https://github.com/Makuna/NeoPixelBus/wiki/HtmlColor-object-API

Thank you.

**Comment by Makuna** [comment]
No, those string names are stored in PROGMEM.  So, they can't be read like a normal char* on most platforms.  They are meant for producing a table to compare to incoming html data that has the names in them rather than be used for constants. (see below for more information)

There is no table or list of constants color in this library.  Names are only standard when applied by a governing group, like Html.  So there could be a table with constant names within that context, and even used for that table.  But this has never been request.  You could create an issue for this request.

Normally, you just build a list of the colors you like and use in your sketch.  like...

```
const RgbColor c_Fuschia(255,0,255);  // using decimal
const RgbColor c_Olive(0x80, 0x80, 0); // using hexadecimal
const RgbColor c_BlanchedAlmond(0xffebcd); // I believe this also works, but you may have to do HtmlColor(0xffebcd) first?
```


_Background:_
[This method is used to parse incoming Html data string here](https://github.com/Makuna/NeoPixelBus/wiki/HtmlColor-object-API#template-typename-t_htmlcolornames-size_t-parseconst-char-name).

It must "parse" the string and compare it to known strings, and that is really what those constants are for.  This is time consuming, and the lookup table uses a non-significant amount of flash space.  So, this is really meant for when you are given a html text data packet to convert rather than use like a constant.  The exposed API to parse will not take a PROGMEM string so they can't be used.

If you had a string with a name in it, you would use it like this using the short/smaller table...
```
HtmlColor color;

color.Parse<HtmlShortColorNames>("fuchsia");  // of course, it wouldn't be a inline like this

strip.SetPixelColor(0, color);
```

**Comment by patfelst** [comment]
ok thanks. I got chatgpt to create a list of constants. Not sure this is the best way to code it, but it works for me.
```c++
const HtmlColor HTML_Fuchsia = 0xFF00FF;
const HtmlColor HTML_Gainsboro = 0xDCDCDC;
const HtmlColor HTML_GhostWhite = 0xF8F8FF;
const HtmlColor HTML_Gold = 0xFFD700;
const HtmlColor HTML_Goldenrod = 0xDAA520;
const HtmlColor HTML_Gray = 0x808080;
const HtmlColor HTML_Grey = 0x808080;
const HtmlColor HTML_Green = 0x008000;
const HtmlColor HTML_GreenYellow = 0xADFF2F;
// ... etc

strip.SetPixelColor(0, RgbColor(HTML_Fuchsia));
strip.Show();
```
I don't think I can easily set a brightness level using this method, would you have a suggestion to do this?

**Reply by Makuna** [reply, marked as answer]
you should just be able to pass the constant like
```
strip.SetPixelColor(0, HTML_Fuchsia);
```

**Reply by patfelst** [reply]
yes that works. I implemented LED brightness by setting the HTLM color name, then call this function just before each strip.Show().
```c++
void LED_Brightness(uint8_t level) {
  RgbColor color;
  for (uint16_t indexPixel = 0; indexPixel < strip.PixelCount(); indexPixel++) {
    color = strip.GetPixelColor<RgbColor>(indexPixel);
    color.Darken(255 - level);
    strip.SetPixelColor(indexPixel, color);
  }
}
```
For example
```c++
  uint8_t led_bright = 15;  // out of 255

  strip.SetPixelColor(0, HTML_Magenta);
  LED_Brightness(led_bright);
  strip.Show();
  delay(300);

  strip.SetPixelColor(0, HTML_DeepSkyBlue);
  LED_Brightness(led_bright);
  strip.Show();
  delay(300);
```

**Reply by Makuna** [reply]
> I implemented LED brightness by setting the HTLM color name, then call this function just before each strip.Show().
> 
If you want overall brightness (luminance) control and/or gamma adjustment, then you should be using [NeoPixelBusLg](https://github.com/Makuna/NeoPixelBus/wiki/NeoPixelBusLg-object) instead of NeoPixelBus.  It has all the same method calls plus a few to adjust brightness.

Gamma is how luminance is "seen" by the human eye.  LEDs are very linear so if you display a strip with brightness levels from o - 255, it will seem the gradient is overly bright at one end and overly dark at the other.  Using gamma will correct this so that what you see will be a smooth transition from dark to bright across the strip.

**Reply by patfelst** [reply]
Thanks, I tried the "Lg" version and it caused a boot loop, I'm using an ESP32-S3. In fact I tried quite a few of the drivers and the only one that didn't cause a boot loop is
`NeoPixelBus<NeoGrbFeature, NeoEsp32LcdX8Ws2812xMethod> strip(1, PixelPin);`

I assumed this was due to the legacy RMT issue most of the WS2812 drivers out there are experiencing and saw there is already an [issue](https://github.com/Makuna/NeoPixelBus/issues/815) for this. I usually use FastLED but even though they've fixed the RMT issue, the LED colours were incorrect for me. Hence I decided to try NeoPixelBus.

EDIT: Just saw in issue 815 you have a fix in the CORE3 branch, I will try this out soon.

Thanks for the Gamma description, I'd never really known what it does.

**Comment by patfelst** [comment]
Branch CORE3 is working on an M5Stack AtomS3 (ESP32-S3) without boot loop using:
`NeoPixelBusLg<NeoGrbFeature, NeoWs2812xMethod> strip(PixelCount, PixelPin);`

Mind you I only have a single RGB LED (PixelCount = 1). And I can use `strip.SetLuminance(led_bright);`, it works really well, thank you.

I wasn't able to get NeoGamma working with the html colors, it says too many overloaded functions:
```c++
// Declared as global
NeoGamma<NeoGammaTableMethod> colorGamma;

// In setup(void)
RgbColor color = colorGamma.Correct(HTML_Red);
strip.SetPixelColor(0, color);
strip.Show();
```
![image](https://github.com/user-attachments/assets/88c6febc-f721-4628-943c-f28510e3b0cc)

**Reply by Makuna** [reply]
You will run into problems with larger number of LEDs with the Core3 RMT, I believe it's around 85ish.  Otherwise, it should be fine.  An IDF fix is in the pipeline for it.

**Comment by patfelst** [comment]
update, I belive this error is because there's no method for correcting type `HtmlColor`. If I replace `HTML_Red` with `RgbColor red` then it works ok. Maybe I need to forget the whole HTML color names thing and just use normal RgbColor variables :)

**Reply by Makuna** [reply]
let the NeoPixelBusLg handle it for you, you won't run into an issue with the gamma.
Gama classes are tuned to be used internally, so they maybe a little rough when you call them.  So you may have to wrap it a RgColor() on the call.
